### Statistical Tests for Experiment 2

In [1]:
import scipy.stats as stats

import pandas as pd
import numpy as np

In [2]:
#Prepare data

def prepare_data(csv_path):
    df = pd.read_csv(csv_path)
    df = df.replace([float('inf'), float('-inf')], float('nan')).dropna()
    # Extract only the numeric value from "Mean±Std" columns
    # Example: '0.549±0.477' -> 0.549
    for col in df.columns:
        if df[col].dtype == 'object' and '±' in str(df[col].iloc[0]):
            df[col] = df[col].apply(lambda x: float(str(x).split('±')[0]))
    return df

In [4]:
# Data Preprocessing

df = prepare_data('../dataset/sets/exp2_sets.csv')

# 2. Re-check for any remaining NaNs or Infs explicitly
df = df.dropna(subset=['Execution_Time', 'Path_Length', 'Optimality_Ratio'])
df = df.replace([np.inf, -np.inf], np.nan).dropna(subset=['Execution_Time'])

display(df)

,Experiment,Heuristic,Movement,Grid_Size,Execution_Time,Nodes_Expanded,Path_Length,Optimality_Ratio
0,2,Manhattan,4-dir,50x50,4.665,428,99,2.020
1,2,Manhattan,4-dir,50x50,5.480,414,99,2.020
2,2,Manhattan,4-dir,50x50,5.430,494,99,2.020
3,2,Manhattan,4-dir,50x50,16.660,1352,101,2.061
4,2,Manhattan,4-dir,50x50,9.500,863,101,2.061
...,...,...,...,...,...,...,...,...
175,2,Chebyshev,8-dir,50x50,8.315,850,58,1.184
176,2,Chebyshev,8-dir,50x50,10.360,975,61,1.245
177,2,Chebyshev,8-dir,50x50,10.940,1055,61,1.245
178,2,Chebyshev,8-dir,50x50,8.565,782,53,1.082


In [5]:
# One-Way ANOVA
groups = [group['Execution_Time'].values for name, group in df.groupby('Heuristic')]
f_stat, p_val = stats.f_oneway(*groups)
print(f"One-Way ANOVA: F={f_stat:.4f}, p={p_val:.4f}")


One-Way ANOVA: F=106.0409, p=0.0000


In [11]:
# Eta Squared (η²) Calculation
ss_total = np.sum((df['Execution_Time'] - df['Execution_Time'].mean())**2)
ss_between = np.sum([len(g) * (g['Execution_Time'].mean() - df['Execution_Time'].mean())**2 for n, g in df.groupby('Heuristic')])
eta_sq = ss_between / ss_total
print(f"Execution Time Effect Size (η²): {eta_sq:.4f}")

Execution Time Effect Size (η²): 0.5451


In [12]:
# Eta Squared (η²) Calculation
ss_total = np.sum((df['Nodes_Expanded'] - df['Nodes_Expanded'].mean())**2)
ss_between = np.sum([len(g) * (g['Nodes_Expanded'].mean() - df['Nodes_Expanded'].mean())**2 for n, g in df.groupby('Heuristic')])
eta_sq = ss_between / ss_total
print(f"Nodes Expanded Effect Size (η²): {eta_sq:.4f}")

Nodes Expanded Effect Size (η²): 0.5872


In [13]:
# Eta Squared (η²) Calculation
ss_total = np.sum((df['Optimality_Ratio'] - df['Optimality_Ratio'].mean())**2)
ss_between = np.sum([len(g) * (g['Optimality_Ratio'].mean() - df['Optimality_Ratio'].mean())**2 for n, g in df.groupby('Heuristic')])
eta_sq = ss_between / ss_total
print(f"Optimality Ratio Effect Size (η²): {eta_sq:.4f}")

Optimality Ratio Effect Size (η²): 0.0027


Statistical Test Summary

One-Way ANOVA: F=106.0409, p=0.0000

Execution Time Effect Size (η²): 0.5451 (strong)

Nodes Expanded Effect Size (η²): 0.5872 (strong)

Optimality Ratio Effect Size (η²): 0.0027 (very weak, we can neglect)